File: Python/day12/main.py
import time
from enum import StrEnum, auto
from functools import lru_cache


class Condition(StrEnum):
    Operational = "."
    Damaged = "#"
    Unknown = "?"


def multiply_data(conditions: list[Condition], blocks: list[int], factor: int) -> tuple[list[Condition], list[int]]:
    new_conditions = conditions.copy()
    new_blocks = blocks.copy()
    for i in range(factor - 1):
        new_conditions.append(Condition.Unknown)
        new_conditions.extend(conditions)
        new_blocks.extend(blocks)
    return new_conditions, new_blocks


class Process:
    def __init__(self, conditions: list[Condition], blocks: list[int]):
        self.conditions = conditions
        self.blocks = blocks

    @lru_cache(maxsize=None)
    def process(self, cond_index: int, block_index: int) -> int:
        if cond_index >= len(self.conditions):
            if block_index >= len(self.blocks):
                return 1
            else:
                return 0
        condition = self.conditions[cond_index]
        if condition == Condition.Operational:
            return self.process(cond_index + 1, block_index)
        if condition == Condition.Damaged:
            if (block_index < len(self.blocks) and
                    self.check_if_block_fits(cond_index, self.blocks[block_index])):
                return self.process(cond_index + self.blocks[block_index] + 1, block_index + 1)
            return 0
        if condition == Condition.Unknown:
            undamaged_sum = self.process(cond_index + 1, block_index)
            damaged_sum = 0
            if block_index < len(self.blocks) and self.check_if_block_fits(cond_index, self.blocks[block_index]):
                damaged_sum = self.process(cond_index + self.blocks[block_index] + 1, block_index + 1)
            return undamaged_sum + damaged_sum

    def check_if_block_fits(self, cond_index: int, block: int) -> bool:
        current_cond_index = cond_index
        current_block = block
        while (current_cond_index < len(self.conditions) and current_block > 0 and
               (self.conditions[current_cond_index] == Condition.Damaged or
                self.conditions[current_cond_index] == Condition.Unknown)):
            current_cond_index += 1
            current_block -= 1
        return (current_block == 0 and
                (current_cond_index == len(self.conditions) or
                 self.conditions[current_cond_index] == Condition.Operational or
                 self.conditions[current_cond_index] == Condition.Unknown))


def conditions_to_str(conditions: list[Condition]) -> str:
    return "".join([c.value for c in conditions])


def load_data(filename: str) -> list[tuple[list[Condition], list[int]]]:
    data = []
    with open(filename, 'r') as file:
        for line in file:
            conditions, values = line.strip().split(' ')
            conditions = [Condition(condition.strip()) for condition in conditions]
            values = [int(value) for value in values.split(',')]
            data.append((conditions, values))
    return data


def solution(factor: int):
    data = load_data("input.txt")
    total_sum = 0
    for conditions, blocks in data:
        conditions, blocks = multiply_data(conditions, blocks, factor)
        process = Process(conditions, blocks)
        total_sum += process.process(0, 0)
    print(total_sum)


def solution_part_1():
    solution(1)


def solution_part_2():
    start = time.time_ns()
    solution(5)
    end = time.time_ns()
    print(f"Time: {(end - start)//1_000_000}ms")


def main():
    solution_part_1()
    solution_part_2()


if __name__ == "__main__":
    main()

File: Rust/day12/src/main.rs

#[derive(Debug, Clone, PartialEq, Eq)]
enum Condition {
    Operational,
    Damaged,
    Unknown
}

struct Process {
    conditions: Vec<Condition>,
    blocks: Vec<usize>,
    cache: Vec<i64>,
    cache_hits: i64,
    invocations: i64
}

const GRID_OVERSIZE: usize = 2;

impl Process {
    fn wrap_process_condition(&mut self, index: usize, block_index: usize) -> i64 {
        self.invocations += 1;
        let cache_value = self.cache[index * (self.blocks.len() + GRID_OVERSIZE) + block_index];
        if cache_value != -1 {
            self.cache_hits += 1;
        } else {
            let result = self.process_condition(index, block_index);
            self.cache[index * (self.blocks.len() + GRID_OVERSIZE) + block_index] = result;
        }
        self.cache[index * (self.blocks.len() + GRID_OVERSIZE) + block_index]
    }

    fn process_condition(&mut self, index: usize, block_index: usize) -> i64 {
        if index >= self.conditions.len() {
            return if block_index >= self.blocks.len() {
                1
            } else {
                0
            }
        }
        let current_condition = &self.conditions[index];
        match current_condition {
            Condition::Unknown => {
                let non_damaged_sum = self.wrap_process_condition(index + 1, block_index);
                let mut damaged_sum = 0;
                if block_index < self.blocks.len() && self.check_if_block_fits(self.blocks[block_index], index) {
                    damaged_sum = self.wrap_process_condition(index + self.blocks[block_index] + 1, block_index + 1);
                }
                return non_damaged_sum + damaged_sum;
            }
            Condition::Operational => {
                return self.wrap_process_condition(index + 1, block_index);
            }
            Condition::Damaged => {
                if block_index < self.blocks.len() && self.check_if_block_fits(self.blocks[block_index], index) {
                    return self.wrap_process_condition(index + self.blocks[block_index] + 1, block_index + 1);
                }
            }
        }
        0
    }

    fn check_if_block_fits(&self, block: usize, index: usize) -> bool {
        let mut current_index = index;
        let mut current_block = block;
        while current_index < self.conditions.len() && current_block > 0 &&
            (self.conditions[current_index] == Condition::Unknown || self.conditions[current_index] == Condition::Damaged) {
            current_index += 1;
            current_block -= 1;
        }
        current_block == 0 && (current_index == self.conditions.len() ||
            (self.conditions[current_index] == Condition::Unknown || self.conditions[current_index] == Condition::Operational))
    }

}

fn load_data(filename: &str) -> Vec<(Vec<Condition>, Vec<usize>)> {
    let data = std::fs::read_to_string(filename).expect("Error reading input file!");
    let mut result = Vec::new();
    for line in data.lines() {
        let mut conditions = Vec::new();
        let mut line_split = line.split(' ');
        for char in line_split.next().unwrap().chars() {
            match char {
                '.' => conditions.push(Condition::Operational),
                '#' => conditions.push(Condition::Damaged),
                '?' => conditions.push(Condition::Unknown),
                _ => panic!("Invalid condition!")
            }
        }
        let values: Vec<usize> = line_split.next().unwrap().trim().split(',').map(|x| x.parse().unwrap()).collect();
        result.push((conditions, values));
    }
    result
}
fn multiply_data(conditions: Vec<Condition>, values: Vec<usize>, factor: usize) -> (Vec<Condition>, Vec<usize>) {
    let mut result_conditions = conditions.to_vec();
    let mut result_values = values.to_vec();
    for _i in 0..(factor - 1) {
        result_conditions.push(Condition::Unknown);
        result_conditions.extend(conditions.to_vec());
        result_values.extend(values.to_vec());
    }
    (result_conditions, result_values)
}

fn solution(factor: usize) {
    let data = load_data("input.txt");

    let mut sum: i64 = 0;
    let mut cache_hits: i64 = 0;
    let mut invocations: i64 = 0;
    for (conditions, values) in data[5..].iter().cloned() {
        let (conditions, values) = multiply_data(conditions, values, factor);
        let cache = vec![-1; (conditions.len() + GRID_OVERSIZE) * (values.len() + GRID_OVERSIZE)];
        let mut process = Process { conditions, blocks: values, cache, cache_hits: 0, invocations: 0};
        sum += process.process_condition(0, 0);
        cache_hits += process.cache_hits;
        invocations += process.invocations;
    }
    println!("{}", sum);
    println!("Cache hits: {}", cache_hits);
    println!("Invocations: {}", invocations);

}

fn solution_part_1() {
    solution(1);
}

fn solution_part_2() {
    solution(5);
}

fn main() {
    solution_part_1();
    solution_part_2();
}
